### DISCUSSION TITLE
ProtoField for abstract class

### DISCUSSION THREAD

**Discussion by srpn** [infinispan/protostream#259] [Q&A]
I am trying to create a `@ProtoAdapter` for a third party class and I couldn't find examples for usage with inheritance.
Lets say there is a class called `CachedAnimal` that contains an `Animal` and `CacheDuration`. 
```
@ProtoAdapter(CachedAnimal.class)
public class CachedAnimalAdapter {
  @ProtoFactory 
  CachedAnimal create(Animal animal, String cacheDuration) {
    return new CachedAnimal(animal, cacheDuration);
  }

  @ProtoField
  Animal getAnimal(CachedAnimal cachedAnimal) {
     cachedAnimal.getAnimal();
  }
  
  @ProtoField
  String getCacheDuration(CachedAnimal cachedAnimal) {
     cachedAnimal.getDuration();
  }
}
```

However, this fails to compile with `@AutoProtoSchemaBuilder` saying the `@ProtoField` type "should not be abstract".

Please provide details on how to deal with abstract types and add some examples to the  documentation.

**Comment by fax4ever** [comment]
Hi @srpn,
Exactly the idea of the `ProtoAdapter` is to generate the marshallers and the schema for class we cannot modify.
I think that you need to set the number attribute in the `@ProtoField` annotation.

**Comment by fax4ever** [comment]
Moreover, you should annotate the class `Animal` with the `@ProtoField`s or define a ProtoAdapter for it in turn.
Can you share more about `Animal` please?

**Comment by srpn** [comment]
Thanks for the response @fax4ever,
Yeah, the main question is how to add annotations to `Animal` and its concrete subclasses. As I mentioned, the above example with `@AutoProtoSchemaBuilder` fails to compile saying the @ProtoField type "should not be abstract".
```
abstract class Animal {
  String type;
}

class Dog extends Animal {
  String name;
  void bark();
  // ...
}
```

**Comment by fax4ever** [comment]
What would be the Protocol Buffers schema you would like to generate for this case?

**Comment by srpn** [comment]
I am looking to use Infinispan for a distributed cache and I am new to Protocol Buffers. 
From what I read, `oneOf` could be a way to represent this as a protobuf message

Here is a full example for ease of reproducing
`
error: org.infinispan.protostream.annotations.ProtoSchemaBuilderException: The type com.test.Animal of field 'animal' of com.test.CachedAnimalAdapter should not be abstract.
`
Code used:
```
package com.test;

import org.infinispan.protostream.annotations.ProtoAdapter;
import org.infinispan.protostream.annotations.ProtoFactory;
import org.infinispan.protostream.annotations.ProtoField;

@ProtoAdapter(CachedAnimal.class)
public class CachedAnimalAdapter {
    @ProtoFactory
    CachedAnimal create(Animal animal, String cacheDuration) {
        return new CachedAnimal(animal, cacheDuration);
    }

    @ProtoField(number = 1)
    Animal getAnimal(CachedAnimal cachedAnimal) {
        return cachedAnimal.getAnimal();
    }

    @ProtoField(number = 2)
    String getCacheDuration(CachedAnimal cachedAnimal) {
        return cachedAnimal.getCachedTime();
    }
}

// Let's say the below classes are from a 3rd party library that I cannot annotate
class CachedAnimal {
    private Animal animal;
    private String cachedTime;

    public CachedAnimal(Animal animal, String cachedTime) {
        this.animal = animal;
        this.cachedTime = cachedTime;
    }

    public Animal getAnimal() {
        return animal;
    }

    public String getCachedTime() {
        return cachedTime;
    }
}

abstract class Animal {
    String type;

    public Animal(String type) {
        this.type = type;
    }
}

class Dog extends Animal {
    private String breed;

    public Dog(String type, String breed) {
        super(type);
        this.breed = breed;
    }

    public void makeSound() {
        System.out.println("woof");
    }
}

class Cat extends Animal {
    private String name;

    public Cat(String type, String name) {
        super(type);
        this.name = name;
    }

    public void makeSound() {
        System.out.println("meow");
    }
}
```

**Comment by fax4ever** [comment]
Yes! One-of seems to be the proper way to map the polymorphism.
At the moment I think that we don't support the schema generation for this case with `@AutoProtoSchemaBuilder`.
I'm opening an issue for it to the ProtoStream Jira => https://issues.redhat.com/projects/IPROTO.
In the meantime I think you could generate the mashallers and the proto schema manually, as we used to do.

**Comment by fax4ever** [comment]
created https://issues.redhat.com/browse/IPROTO-267
